### PR TITLE
[FLINK-4260] Support specifying ESCAPE character in LIKE and SIMILAR TO expression.

### PR DESCRIPTION
**Pull request by miaoever**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

…ions.

### CODE REVIEW

**Comment by miaoever** [comment]
@twalthr @wuchong Would you please review my pr to flink, thx :)

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java [hunk 1]
```diff
@@ -118,4 +119,39 @@ public void testJoin() throws Exception {
 		String expected = "Hi,Hallo\n" + "Hello,Hallo Welt\n" + "Hello world,Hallo Welt\n";
 		compareResultAsText(results, expected);
 	}
+
+
+	@Test
+	public void testLikeWithEscapeFromDataSet() throws Exception {
```
Comment (wuchong) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java]: We try to keep the number of integration tests for the Table API to a minimum to not exceed Travis' build time threshold of 2 hours. I think we do not need the  IT cases, the `ScalarFunctionsTest` should cover all cases. 

You can add new tests to `ScalarFunctionsTest` called `testLikeWithEscape`, `testSimilarWithEscape` just like what `testLike` did.

Comment (wuchong) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java]: Oh, and addd `testNotLikeWithEscape ` , `testNotSimilarWithEscape`  cases. 

I'm not sure whether we should support LIKE ESCAPE for Table API. cc @fhueske  @twalthr  what do you think ?

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java [hunk 2]
```diff
@@ -118,4 +119,39 @@ public void testJoin() throws Exception {
...
+		Table result = tableEnv.sql(sqlQuery);
+
+		DataSet<Row> resultSet = tableEnv.toDataSet(result, Row.class);
+		List<Row> results = resultSet.collect();
+		String expected = "%leo*\n" + "%deng*\n" + "%hello*\n";;
+		compareResultAsText(results, expected);
+	}
+
+	@Test
+	public void testSimilarWithEscapeFromDataSet() throws Exception {
```
Comment (wuchong) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java]: remove this too.

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/util/CollectionDataSets.java [hunk 3]
```diff
@@ -54,6 +51,18 @@
  */
 public class CollectionDataSets {
 
+	public static DataSet<Tuple1<String>> getTupleDateSet(ExecutionEnvironment env) {
+
+		List<Tuple1<String>> data = new ArrayList<>();
+		data.add(new Tuple1<String>("%leo*"));
+		data.add(new Tuple1<String>(("%deng*")));
+		data.add(new Tuple1<String>("%hello*"));
```
Comment (wuchong) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/util/CollectionDataSets.java]: I think we should add some negative test data (without `%` and `*`).

And I would like not to touch other modules if possible, it's better to move the specific test data to the specific test class.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java (hunk 1 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java]: IMHO we don't need it for the Table API. There are very few use cases for it.

Comment (wuchong) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/sql/SqlITCase.java]: OK

**Comment by miaoever** [comment]
Hi @wuchong , I have modified the test cases according to your comments, please review my latest commit. Thanks :)

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/util/CollectionDataSets.java [hunk 4]
```diff
@@ -33,10 +33,7 @@
 
 import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
 import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo;
-import org.apache.flink.api.java.tuple.Tuple2;
-import org.apache.flink.api.java.tuple.Tuple3;
-import org.apache.flink.api.java.tuple.Tuple5;
-import org.apache.flink.api.java.tuple.Tuple7;
+import org.apache.flink.api.java.tuple.*;
```
Comment (wuchong) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/util/CollectionDataSets.java]: Could you revert this change ? We would like to avoid `.*` imports.

**Comment by wuchong** [comment]
It would be great if you can rebase your branch and resolve the conflict.

**Comment by miaoever** [comment]
Hi @wuchong, I have resolved the conficts :)

**Comment by wuchong** [comment]
The Travis CI has failed because of  `BuiltInMethods` line length, could you fix this ? 
```
[INFO] --- scalastyle-maven-plugin:0.8.0:check (default) @ flink-table_2.10 ---
error file=/home/travis/build/apache/flink/flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/BuiltInMethods.scala message=File line length exceeds 100 characters line=35
error file=/home/travis/build/apache/flink/flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/BuiltInMethods.scala message=File line length exceeds 100 characters line=36
Saving to outputFile=/home/travis/build/apache/flink/flink-libraries/flink-table/target/scalastyle-output.xml
Processed 229 file(s)
```

**Comment by miaoever** [comment]
@wuchong  sure, fixed :)

**Comment by wuchong** [comment]
Great ! The PR looks good to me now, thank you for your contribution !

+1

**Comment by twalthr** [comment]
The PR looks good. The only thing missing is documentation. I will add it and merge this.
